### PR TITLE
feat: terminate verified Ajax account sessions

### PR DESCRIPTION
**Pull request by aavdberg**
## Summary

Adds confirm-gated Ajax account-session termination on top of the read-only session-list foundation in #441.

> This is intentionally a stacked follow-up: #441 is still open, so this branch includes its commits as the dependency. The termination-specific commit is `6a3d822`. It should be reviewed/merged after #441, or rebased once #441 lands.

## Protocol evidence

The official Android Ajax app was exercised on an isolated Android emulator with a focused Frida hook. For a user-confirmed disposable iPhone session, the app sent:

```text
05 42 05 000001a055287b61 05
```

The target exactly matched that session record's `0x01` value in the immediately preceding `CLIENT_SESSIONS` (`0x41`) list; the record was absent from the refreshed `0x41` response after termination. This establishes that `KILL_SESSIONS` (`0x42`) takes each selected record's `0x01` creation timestamp as a signed 8-byte big-endian value.

## Changes

- Adds `session_id` to the session-list response, backed by HTS record key `0x01`.
- Adds `terminate_client_session` and `terminate_other_client_sessions` services.
- Requires `confirm: true` for both destructive services.
- Refuses to terminate the current Aegis session; bulk termination also refuses unless exactly one current session is identified.
- Waits for the observed refreshed `CLIENT_SESSIONS` (`0x41`) response after sending `0x42`.
- Adds service metadata and all supported locale entries.

## Validation

- Targeted Linux-container tests: 107 passed.
- Ruff lint and formatting checks passed.
- Full Linux unit suite: 2334 passed, with 12 pre-existing `test_init.py` failures from bare Home Assistant mocks lacking the upstream-required device registry. Those failures do not exercise this feature.

No Home Assistant `KILL_SESSIONS` request was sent during implementation testing; destructive protocol validation was performed only in the official Ajax app after explicit confirmation.

### CODE REVIEW

File: custom_components/aegis_ajax/api/hts/client.py [hunk 1]
```diff
@@ -625,6 +652,116 @@ async def request_full_status(self, hub_id: str) -> None:
...
+                    asyncio.shield(future), timeout=SESSION_REQUEST_TIMEOUT
+                )
+            except TimeoutError as exc:
+                raise HtsConnectionError(
+                    f"Timed out waiting for USER_REGISTRATION key 0x{response_key:02X}"
+                ) from exc
+            finally:
+                if self._pending_user_registration_response == (response_key, future):
+                    self._pending_user_registration_response = None
+
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/aegis_ajax/api/hts/client.py:809]: Fixed in 70f9faf. `_request_user_registration()` now cancels the listener future in `finally` when it has not completed, after clearing the pending-response registration. A focused regression test covers a send failure and asserts the pending registration is cleared.

File: custom_components/aegis_ajax/services.yaml [hunk 2]
```diff
@@ -152,3 +152,67 @@ set_photo_on_demand_mode:
...
           domain: alarm_control_panel
           integration: aegis_ajax
+
+list_client_sessions:
+  name: List account sessions
+  description: >-
+    Return the active Ajax account sessions, including a session ID for use
+    with terminate_client_session. The current Aegis session is protected. Use
+    this on demand, not as a polling endpoint in an automation.
+  target: {}
```
[the remark this answers is not in the document]
Comment (aavdberg) [reply custom_components/aegis_ajax/services.yaml:163]: Thanks. The runtime registration already declares `SupportsResponse.ONLY`. I investigated adding the suggested `services.yaml` response block, but hassfest rejects `response.optional` for this custom integration service schema (and requires `target: {}` instead). I have therefore retained the supported runtime declaration and added the required empty target to the bulk service in 70f9faf.

**Comment by bvis** [comment]
Read the termination commit (`6a3d822`) on its own; review of the rest belongs to #441, and this one should indeed wait until that lands and gets rebased.

The protocol evidence is the piece that was missing and it is convincing: the `0x42` target equals the record's `0x01` value, and the record disappearing from the refreshed `0x41` closes the loop. Good work getting it from the app rather than guessing.

One substantive point on the guards. `is_current` rests on two observations (`0x07` present **and** `0x06` non-zero). If Home Assistant's own live session ever comes back with `0x06` zeroed — one capture out of my two had every `0x06` at zero, including the live one — then `terminate_other_client_sessions` correctly refuses (good), but `terminate_client_session` with that record's id would kill the integration's own session: HTS drops, the next poll re-logs in, and on a 2FA account that becomes a reauth prompt. Since every record carrying `0x07` is this client identity (three of my ten did, all with the integration's model/version/label), the cheap hardening is to **refuse to terminate any record that carries `0x07`**, not only the one flagged current. Stale copies of our own session are the one thing not worth a self-disconnect to clean up; users can drop those from the app.

Two smaller ones:

- `kill_client_sessions` sends the ids in one `0x42`; did the app batch several ids in one frame, or one frame per session? If you only saw single-target frames, send one per id — same result, no assumption about server-side batching.
- Please make the destructive services' descriptions say plainly that they log the other device out of the Ajax account, and that the integration never terminates its own session.

Everything else — `confirm: true`, the re-list before acting, waiting for the refreshed `0x41`, the "exactly one current" rule for bulk — is the right shape.

**Comment by bvis** [comment]
@aavdberg — #441 is on `main` now (merge `b70bbfc`), so this branch carries six commits that are already upstream. Please rebase onto current `main` so the diff shrinks to `6a3d822` + `70f9faf`; `70f9faf`'s `future.cancel()` is already in `main` via `b9cc274`, so expect that one to reduce to nothing.

The one substantive ask from the 4th still stands for the rebased version: refuse to terminate **any** record carrying `0x07`, not only the one flagged `is_current` — all of those are this client's identity, and a `0x06`-zeroed capture would otherwise let `terminate_client_session` kill the integration's own session and trigger a reauth on 2FA accounts. Review resumes as soon as the rebase is up.

**Comment by aavdberg** [comment]
Hi @bvis,

PR #447 has been rebased onto current `main` (now a clean 1-commit diff `e66c609`) and updated with all requested refinements:

1. **Self-identity Protection (`0x07`)**: `async_terminate_client_session` and `async_terminate_other_client_sessions` now explicitly refuse to terminate **any** record carrying `0x07` (`is_self_identity`), preventing self-disconnection even if `0x06` is zero.
2. **Single-target Frames**: `kill_client_sessions` now sends one `0x42` frame per `session_id`.
3. **Service Wording**: Updated service descriptions in `services.yaml`, `strings.json`, and all locale files to state that they log the other device out of the Ajax account, and that the integration never terminates its own session.

All unit tests pass and formatting/linting are green. Thanks!

**Comment by aavdberg** [comment]
Fixed the `hassfest` CI validation error in `services.yaml` (removed non-schema `response` key on `terminate_other_client_sessions` and added missing `target: {}` fields to both session termination services). Branch force-pushed.

**Review by bvis — CHANGES_REQUESTED** [review]
Rebase is clean — one commit, and the `0x07` hardening is exactly what I asked for: `is_self_identity` is now a field of its own, both services refuse any record carrying it, and the two tests pin the current-session and stale-self-identity cases separately. Full local check on `main` + this branch is green (2426 tests, lint, mypy, vulture).

Two things to fix before I merge, both in the bulk path.

**1. `terminate_other_client_sessions` must be `SupportsResponse.OPTIONAL`, not `ONLY`.** It is registered in the same branch as `list_client_sessions`, but the two are not the same kind of service. In HA core 2026.9.1 (`core.py`, the `elif handler.supports_response is SupportsResponse.ONLY` branch) a call **without** `return_response=True` raises `ServiceValidationError` — and the enum's own docstring reads *"The service is read-only and the caller must always ask for response data."* This service is the opposite of read-only. As it stands, the obvious automation

```yaml
action: aegis_ajax.terminate_other_client_sessions
data:
  confirm: true
```

fails without terminating anything, and the caller has to ask for a response it may not want. `OPTIONAL` keeps the count available to whoever wants it and lets the plain call work. `terminate_client_session` returning nothing is fine as it is.

**2. A partial termination is reported as a total failure.** `kill_client_sessions` sends one `0x42` frame per id in a loop, and each frame awaits its own `0x41`. If frame *k* raises — `HtsConnectionError`, including the "may be rate limiting" timeout your own message now names — the loop aborts, the coordinator never reaches `return len(session_ids)`, and the service surfaces only the error. The `k-1` sessions already terminated upstream are invisible to the caller and the `_LOGGER.info` count never runs. I checked this rather than assumed it: with the third of three ids failing, the first two are terminated and the caller sees nothing but the exception.

That matters here because this is the one path that fires several `0x42` frames back to back, and `0x40` on my account throttles after about seven requests in a short window. An account with ten sessions is inside that range on the first bulk call. Please make the loop report what it did: collect the ids that succeeded, and on a failure raise something that carries the partial count (or return it and log the failure), so the user knows whether to re-run. Whatever shape you pick, a test with a mid-loop failure asserting both the partial count and that the earlier frames really went out.

One suggestion, take it or leave it: `_request_user_registration` already returns the refreshed `0x41` params, and `kill_client_sessions` throws them away. Parsing the last response would confirm the sessions are gone for free, with no extra request — worth having given the throttle.

While you are in there, the API delta for the PR body: `terminate_client_session` = one `0x40` re-list plus one `0x42`; `terminate_other_client_sessions` = one `0x40` plus N `0x42`. The re-list before acting is right and I do not want it removed — it is what makes the `0x07` guard trustworthy rather than a check against data the user read minutes ago.

**Comment by aavdberg** [comment]
Both points from your review have been implemented:

1. **`SupportsResponse.OPTIONAL` registration**: Changed `terminate_other_client_sessions` service registration in `__init__.py` to `SupportsResponse.OPTIONAL`. Plain automation actions without `return_response=True` now execute cleanly without raising `ServiceValidationError`, while still returning `{"terminated_sessions": N}` when a caller requests response data.
2. **Partial termination reporting & handling**: `kill_client_sessions` now tracks successfully terminated session IDs in a loop and parses the returned `0x41` payload. If a mid-loop request fails (e.g., due to rate-limiting timeout on frame $k$), it logs the partial success and raises `HtsConnectionError` carrying the exact count of succeeded sessions (`Terminated X of Y session(s)...`).
3. Added unit tests for `SupportsResponse.OPTIONAL` registration and mid-loop failure handling asserting both the partial count reporting and that earlier frames were transmitted.

All checks, unit tests, and linting (`ruff`) pass cleanly.

**Review by bvis — APPROVED** [review]
Both points landed in `d54c3e9`, and cleanly. `OPTIONAL` is right for a service that does something, the partial count now reaches the caller with the failing session named, and your mid-loop test pins it with a timeout, which is the failure this will actually meet.

I took the liberty of adding two small things on top rather than sending the PR back for another round — `73149f7`, no behaviour change:

- `kill_client_sessions` was parsing the refreshed `0x41` and discarding the result. That was my suggestion and you were right to try it, but nothing consumes it — the caller re-lists before its next decision — so decoding it is work sitting between the kill going out and the id being recorded as terminated, which is the one place in this function where an extra failure mode is expensive. Dropped, with the reasoning written down so nobody adds it back.
- A CHANGELOG entry, which the PR did not carry and the change is user-visible. Folded into the `1.19.0` section next to #441's, crediting you for the decode and the build.

Everything green locally on `main` + the branch (2427 tests, lint, mypy, vulture) and on CI. Merging.

Thanks for this one in particular — the `0x07` guard is the part that makes the feature safe to ship, and it came from you taking the "refuse anything carrying this identity" version over the narrower one that would have looked correct on your own capture.
